lib/multiplexer/proxyline.go

```diff
@@ -238,11 +290,66 @@ func ReadProxyLineV2(reader *bufio.Reader) (*ProxyLine, error) {
 		return nil, trace.BadParameter("unsupported protocol %x", header.Protocol)
 	}
 
+	// If there are more bytes left it means we've got TLVs
 	if header.Length > size {
-		if _, err := io.CopyN(io.Discard, reader, int64(header.Length-size)); err != nil {
+		tlvsBytes := &bytes.Buffer{}
+
+		if _, err := io.CopyN(tlvsBytes, reader, int64(header.Length-size)); err != nil {
 			return nil, trace.Wrap(err)
 		}
+
+		tlvs, err := UnmarshalTLVs(tlvsBytes.Bytes())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		ret.TLVs = tlvs
 	}
 
 	return &ret, nil
 }
+
+// UnmarshalTLVs parses provided bytes slice into slice of TLVs
+func UnmarshalTLVs(rawBytes []byte) ([]TLV, error) {
+	var tlvs []TLV
+	for i := 0; i < len(rawBytes); {
```

Comment:
I wonder if we can change this to `for len(rawBytes)>0` and then just shorten slice instead of modifying `i`.
```go
var tlvs []TLV
for len(rawBytes) > 0 {
	if len(rawBytes) < 3 {
		return nil, ErrTruncatedTLV
	}
	tlv := TLV{
		Type: PP2Type(rawBytes[0]), // First byte is TLV type
	}
	// Next two bytes are TLV's value length
	tlvLen := int(binary.BigEndian.Uint16(rawBytes[1:]))
	rawBytes = rawBytes[3:] //skip header
	if tlvLen > len(rawBytes) {
		return nil, ErrTruncatedTLV
	}
	// Ignore no-op padding
	if tlv.Type != PP2TypeNOOP {
		tlv.Value = make([]byte, tlvLen)
		copy(tlv.Value, rawBytes[:tlvLen]) // I think even copy(tlv.Value, rawBytes) would do
	}
	rawBytes = rawBytes[tlvLen:]
	tlvs = append(tlvs, tlv)
}
return tlvs, nil
```

Reply:
Applied 👍